### DISCUSSION TITLE
Proposals: On adding Gazebo Harmonic with ```gz_ros2_control``` support

### DISCUSSION THREAD

**Discussion by Mechazo11** [space-ros/space-ros#242] [Ideas]
Dear Space-ROS team,

Apologies in advance, this turned out to be a longer post than I had anticipated.

Recently I ported an UGV [simulator](https://github.com/Mechazo11/clearpath_simulator_harmonic) from Ignition Fortress to Gazebo Harmonic using source built versions of ROS 2 Jazzy, Gazebo Harmonic, Moveit2 and Nav2 in Ubuntu 22.04. Some of my major observations are as follows 

* Gazebo Harmonic [packages](https://github.com/Mechazo11/gazebo_harmonic_ws/blob/main/collection-harmonic.yaml) may not directly depend upon ROS 2 jazzy compliant packages **but** requires a number of ```vendor``` packages to interface with different third party software (i.e. OGRE 2). A (good) thing about source building Harmonic, once the build is done, the ```build``` directory is no longer required and can be deleted to recover a good amount of disk space. 

* ```gz_ros2_control``` is a mandatory requirement to utilize any of the ```ros2_controllers```. ```ros2_controllers``` in turn depends upon ```ros2_control``` and other pertinent packages to work. 

* I experimented with the idea of combining **all** Harmonic packages and their corresponding ```vendors``` package into one workspace. However it prevented ```gz server``` from properly launching any world files regardless if the configurations were done correctly. Whats more, this problem throws no run time error only hangs up the gz server into an infinite loop stating ```looking for world files```.

Given the above, I propose the following for integrating Gazebo Harmonic

A. Starting from the ```jazzy``` branch, Space-ROS should only support Gazebo Harmonic (recommended LTS version with ROS 2 Jazzy) and remove/discourage any and all usage of Ignition Fortress.

B. To integrate Harmonic, I think it is strategically beneficial to have a variable / flag in the Earthly script to build the base image with an additional workspace called ```harmonic_ws``` where only the constituent packages of Harmonic are built. In space-ros ```humble``` branch, once all packages were built for the base workspace, the ```build``` folder was deleted to reduce image space. The same is applicable for Gazebo Harmonic since the official instructions ```merges``` all built files in the ```install``` folder.

C. The `base + harmonic` **base** image now should be followed by say an ```interface_ws``` where packages such as ```ros2_control```, ```ros2_controller```, ```rviz``` and their dependencies, most if not all the ```vendor``` packages and ```gz_ros2_control``` are built. The resulting image now will be ready for building the ```dev_ws``` where users can simulate space robots.

Thank you for your time. 

With best regards,
@Mechazo11

**Comment by asimonov** [comment]
@Mechazo11 thanks for starting this discussion!

> Recently I ported an UGV [simulator](https://github.com/Mechazo11/clearpath_simulator_harmonic) from Ignition Fortress to Gazebo Harmonic using source built versions of ROS 2 Jazzy, Gazebo Harmonic, Moveit2 and Nav2 in Ubuntu 22.04. Some of my major observations are as follows
> 

I am not sure it is good idea to build this combination on Ubuntu22.04 (and not on officially supported Ubuntu24.04)... I could not pull it off. But if this is possible then NASA may be interested as they are still on Ubuntu22 :)...


> * `gz_ros2_control` is a mandatory requirement to utilize any of the `ros2_controllers`. `ros2_controllers` in turn depends upon `ros2_control` and other pertinent packages to work.

We briefly touched on the requirement to have `ros2_control` in space-ros images before. From what I remember we dont really need it for mars rover demo. Not too sure about moveit2-based demos... I would say we definitely dont want it in base space-ros image. But may be required in some demos...

> Given the above, I propose the following for integrating Gazebo Harmonic
> 
> A. Starting from the `jazzy` branch, Space-ROS should only support Gazebo Harmonic (recommended LTS version with ROS 2 Jazzy) and remove/discourage any and all usage of Ignition Fortress.
> 

YES! Lets move to Ubuntu24.04 + Jazzy + Harmonic as main stack and get rid of Ignition.

> B. To integrate Harmonic, I think it is strategically beneficial to have a variable / flag in the Earthly script to build the base image with an additional workspace called `harmonic_ws` where only the constituent packages of Harmonic are built. In space-ros `humble` branch, once all packages were built for the base workspace, the `build` folder was deleted to reduce image space. The same is applicable for Gazebo Harmonic since the official instructions `merges` all built files in the `install` folder.
> 

I suggest we dont add any Gazebo packages into base space-ros image.

Better put separate Gazebo layer in demos docker images.

Also, I am not sure it is good idea to build Gazebo from source. Not sure if it is easy nowdays, but in the past Gazebo build was "complicated". If the source build works reliably and is reasonably fast then it would be nice to have this to fix the versions for space-ros (repeatability is our core principle). 

But we may just be ok to live with apt install, which will be less repeatable if we need to rebuild same image later...

> C. The `base + harmonic` **base** image now should be followed by say an `interface_ws` where packages such as `ros2_control`, `ros2_controller`, `rviz` and their dependencies, most if not all the `vendor` packages and `gz_ros2_control` are built. The resulting image now will be ready for building the `dev_ws` where users can simulate space robots.
> 

I think it is good idea to have 'dev' image that builds on top of 'base' and adds Gazebo and tools like rqt and rviz2, with the dependencies. May be it is good idea to maintain it in base space-ros repo, not in 'demos'...

What do others think?

**Comment by Mechazo11** [comment]
Hi @asimonov 
I was able to get ROS 2 Jazzy to work in Ubuntu 22.04 using a ```custom``` package roster that took me a while to figure out. You can give my Jazzy workspace in Ubuntu 22.04 a try here: https://github.com/Mechazo11/ubuntu22_jazzy_ws. It takes about 45 mins in my machine to build. The package roster was meant to easily integrate Harmonic, Nav2 and Moveit2 hence it is about 400+ packages.

> But we may just be ok to live with apt install, which will be less repeatable if we need to rebuild same image later...

When I was attempting to integrate the O3DE engine, I opened this [ticket](https://github.com/space-ros/space-ros/issues/215). At that time, I was relying on apt-installing a number of ```ROS 2 Humble``` packages which I was told is [considered a bug](https://github.com/space-ros/docker/issues/165). If the ```apt-install``` route is taken, wouldn't that be also considered a bug?

> I think it is good idea to have 'dev' image that builds on top of 'base' and adds Gazebo and tools like rqt and rviz2, with the dependencies. May be it is good idea to maintain it in base space-ros repo, not in 'demos'...

I agree but given Harmonic would be primary simulator engine, I was thinking of having two ```base``` images, one being the ```core``` and the other being ```base + harmonic``` but I guess as long as the order of workspace being built and sourced is correct, it probably doesn't warrant to have two ```base``` images.

**Comment by asimonov** [comment]
I think current versions of demos are on Harmonic. and new GUI stack will have Harmonic gazebo... so probably we can close this
